## 非同步事件&同步事件的差異
### 同步事件： 
- 一次只做一件事情 (TASK1 -> TASK2 -> TASK3)，其他TASK必須等待前一個完成後才能繼續執行，是按順序執行的。
- 優點：執行順序明確
- 缺點：可能會導致阻塞 (Blocking)，因為在等待的過程中，其他任務無法執行
- 範例(參考至:https://developer.mozilla.org/zh-TW/docs/Learn/JavaScript/Asynchronous/Introducing)
```js
const btn = document.querySelector("button");
btn.addEventListener("click", () => {
  alert("你點擊了我！");

  let pElem = document.createElement("p");
  pElem.textContent = "這是一段新加的文字。";
  document.body.appendChild(pElem);
});
```
### 非同步事件：     
- 同時可以做很多件事情，不需要等到前一件事情做完才做下一件事情，例如執行TASK1時，我另一個TASK2也同時在執行，改善了同步事件的缺點
- 特別是在向 API 發送請求時，會很需要用非同步方式，原因是會用到網路，而網路速度是不穩定的，等待資料回傳速度也是不穩定的，因此若是用同步方式，使得user只能空等在那

## Promise then用法
- 簡單來說就是，當Promise達成就做下一件事，反之沒達成就不做下一件事
- 語法：
  ```js
  new Promise(function(resolve, reject) { 
  // if success, resolve()
  // if fail, reject()
  // 要注意Promise的P要大寫
  })
  ```
### 基本概念：
#### 使用 new Promise 來創建一個新的 Promise後，Promise 可以帶入一個函式，而這個函式又會被傳入兩個參數："resolve & reject"。
#### Promise 達成，呼叫 resolve() 來回傳 Promise，接著可以用 "then" 來接續這個非同步執行完後要做的動作。
#### Promise 失敗，呼叫 reject() 來回傳失敗原因，接著可以用 "catch" 來接住錯誤。
#### 當 Promise 被解析成 resolve 或 reject 之後，其值就"無法再被改變"。


